### PR TITLE
Implement a CLI closer to the idealize usage

### PR DESCRIPTION
**Pull request by jessemyers**
Main changes:
 - Dynamically discover available sources
 - Delegate source arguments to source classes
 - Use the `compose` and `init` syntax

### CODE REVIEW

File: docker_etude/main.py [hunk 1]
```diff
@@ -2,39 +2,72 @@
...
-    Generate a composition.
+    Custom group with dynamially generate commands.
 
     """
-    pass
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.add_sources()
+
+    def add_sources(self):
```
Comment (jessemyers) [inline docker_etude/main.py]: By extending the `Group`, we can inject extra commands into the calling chain.

File: docker_etude/main.py [hunk 2]
```diff
@@ -2,39 +2,72 @@
...
+@pass_obj
+def compose(source, modules):
+    """
+    Generate a composition.
+
+    """
+    # always enable the default source last
+    source.sources.append(Source.load_source())
+
+    # XXX still no nice way to pass arguments to the modules
```
Comment (jessemyers) [inline docker_etude/main.py]: The root problem here is that we don't know which modules are available until we select our sources. I haven't found a good way to have `click` re-evaluate its chain part way yet.

File: docker_etude/cli.py [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+
+    Note that click will **ALSO** raise error on `baz` and `--etc` in the context of parsing the
+    `foo` command. While these errors can be suppressed with various context flags, we prefer to catch
+    the errors and truncate our arguments at the first error to avoid the *last* value problem.
+
+    """
+    if isinstance(error, NoSuchOption):
+        targets = [error.option_name]
+    elif isinstance(error, UsageError):
+        match = search("Got unexpected extra arguments? \((.*)\)", error.message)
```
Comment (jessemyers) [inline docker_etude/cli.py]: This regex is the only truly evil thing I had to do to make our desired CLI usage work. The rest of it was tricky, but used well-defined click APIs.

File: docker_etude/cli.py [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+    :param source: a module source
+    :param args: a list of (unparsed) string arguments
+
+    """
+    while full_args:
+        # expect inputs of the form: <module_name> [args] ...
+        module_name, local_args = full_args[0], full_args[1:]
+        module_cls = source.find_module(module_name)
+
+        # generate a command that will create our module
```
Comment (jessemyers) [inline docker_etude/cli.py]: Note that the third argument here is the "callback function", i.e. the function we usually decorate with `@command`. In this case, all I want is to instantiate by module, so the class itself provides the callback function (`__init__`).

File: docker_etude/main.py [hunk 5]
```diff
@@ -2,56 +2,73 @@
...
+class Etude(Group):
     """
-    Generate a composition.
+    Custom CLI group with dynamially generate commands.
 
     """
-    pass
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        # add commands for each available non-default source
```
Comment (jessemyers) [inline docker_etude/main.py]: Trick #1: define a custom `click.Group` that automatically adds a command for each available source. This let us run:

```
etude ecs --region us-east-1
```

File: docker_etude/main.py [hunk 6]
```diff
@@ -2,56 +2,73 @@
...
+        for source_command in iter_source_commands():
+            self.add_command(source_command)
+
+
+@group(chain=True, cls=Etude, invoke_without_command=True)
+@pass_context
+def etude(context):
```
Comment (jessemyers) [inline docker_etude/main.py]: Trick #2: start out the CLI call chain with our custom group and initialize the context to contain `MultiSource`, which will delegate to any sources we specify.

File: docker_etude/main.py [hunk 7]
```diff
@@ -2,56 +2,73 @@
...
+    """
+    Manage a composition.
 
+    """
+    context.obj = MultiSource()
+
+
+@etude.command(name="compose")
```
Comment (jessemyers) [inline docker_etude/main.py]: Trick #3: Define an `UNPROCESSED` argument that captures all remaining arguments once the user types `compose`. This stops processing sources and starts processing modules.

File: docker_etude/main.py [hunk 8]
```diff
@@ -2,56 +2,73 @@
...
-        Source.load_source(
-            source_name="default",
-        ),
-    )
-    # XXX modules should support arguments
+    """
+    # always enable the default source last
+    source.add(Source.load_source())
+
+    # build modules list
```
Comment (jessemyers) [inline docker_etude/main.py]: Trick #4: Parse modules ourself using the unprocessed `extra` arguments.

File: docker_etude/sources/ecs.py [hunk 9]
```diff
@@ -45,30 +52,56 @@ def session(self):
...
-        return ECSExtension.from_ecs(self, cluster, service, module_name)
+        cluster, service = self.service_clusters[module_name]
+
+        kwargs = dict(
+            cluster=cluster,
+            name=module_name,
+            service=service,
+            source=self,
+        )
+        module_cls = type(camelize(module_name), (ECSExtension, ), kwargs)
```
Comment (jessemyers) [inline docker_etude/sources/ecs.py]: Trick #99: Generate a new module class for each ECS service we parse. This allows us to just instantiate all of our modules without having to worry about combining argument from the parser and from the ECS source.

That is:
 - we inject all the arguments from the ECS source into the class defintition
 - we exclusively use the constructor for arguments from the parser

This turns out very cleanly.

File: docker_etude/tests/test_cli.py [hunk 10]
```diff
@@ -0,0 +1,150 @@
...
+            has_properties(
+                name="elasticsearch",
+                cluster_name="example",
+                port=9200,
+            ),
+        ),
+    )
+
+
+def test_iter_modules_multiple_with_overlapping_args():
```
Comment (jessemyers) [inline docker_etude/tests/test_cli.py]: This test was a doozy to make work.

**Comment by cdm-ium** [comment]
+1 I'm sure this magic will be the gift that keeps on giving.

**Pull request closed without merging**